// review.js/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var compilerModel_1 = require("./model/compilerModel");
exports.Book = compilerModel_1.Book;
exports.ContentChunk = compilerModel_1.ContentChunk;
exports.ReportLevel = compilerModel_1.ReportLevel;
exports.ProcessReport = compilerModel_1.ProcessReport;
var controller_1 = require("./controller/controller");
var parser_1 = require("./parser/parser");
exports.SyntaxTree = parser_1.SyntaxTree;
__export(require("./parser/parser"));
var analyzer_1 = require("./parser/analyzer");
exports.AcceptableSyntaxes = analyzer_1.AcceptableSyntaxes;
exports.DefaultAnalyzer = analyzer_1.DefaultAnalyzer;
var validator_1 = require("./parser/validator");
exports.DefaultValidator = validator_1.DefaultValidator;
var builder_1 = require("./builder/builder");
exports.DefaultBuilder = builder_1.DefaultBuilder;
var htmlBuilder_1 = require("./builder/htmlBuilder");
exports.HtmlBuilder = htmlBuilder_1.HtmlBuilder;
var textBuilder_1 = require("./builder/textBuilder");
exports.TextBuilder = textBuilder_1.TextBuilder;
var analyzer_2 = require("./parser/analyzer");
exports.SyntaxType = analyzer_2.SyntaxType;
/**
 * ReVIEW文書のコンパイルを開始する。
 * @param setup
 * @param options
 * @returns {Book}
 */
function start(setup, options) {
    "use strict";
    var controller = new controller_1.Controller(options);
    // setup 中で initConfig が呼び出される
    setup(controller);
    return controller.process();
}
exports.start = start;
